### PR TITLE
WDIP-41: Use WebSockets for message updates

### PR DESCRIPTION
**Pull request by utiq**
## Problem

The messages and overview of the game were being fetched every 2 seconds to keep the UI updated. That was a temporary solution until we could implement something more complex.

[#L162-L163](https://github.com/adamlerer/webDiplomacy/blob/main/beta-src/src/components/ui/WDUI.tsx#L162-L163)
image

## Solution

I was looking for a WebSockets solution that could be flexible. I found the perfect solution for our use case. I implemented [soketi](https://github.com/soketi/soketi). Soketi is an open-source WebSockets server built on top of [uWebSockets.js](https://github.com/uNetworking/uWebSockets.js). The good part of working with Soketi is that we can use a paid solution called [Pusher](https://pusher.com/) or create Soketi instances in any PaaS.

## Troubleshooting

I would leave the console.info for now until the production release is stable, or maybe forever. To verify if the WebSockets subscriptions were successful, just check the console for these lines.

image

### CODE REVIEW

File: config.sample.php [hunk 1]
```diff
@@ -38,7 +38,7 @@ class Config
 	 *
 	 * @var string
 	 */
-	public static $database_socket='mariadb';
+	public static $database_socket = 'mariadb';
```
Comment (adamlerer) [inline config.sample.php]: This looks like more accidental formatting changes.

File: composer.lock [hunk 2]
```diff
@@ -4,7 +4,7 @@
         "Read more about it at https://getcomposer.org/doc/01-basic-usage.md#installing-dependencies",
         "This file is @generated automatically"
     ],
-    "content-hash": "93367a3be337e7b590e4ba9417978c79",
+    "content-hash": "6286d4a3985d893a63324ad5686e38e9",
```
Comment (adamlerer) [inline composer.lock:7]: What is this file?

File: api.php [hunk 3]
```diff
@@ -1173,6 +1173,10 @@ public function run($userID, $permissionIsExplicit) {
 
 		$timeSent = libGameMessage::send($toCountryID, $countryID, $message);
 
+		// I cannot put this require in the first lines of the page, it gives me an error
+		require_once('lib/pusher.php');
```
Comment (adamlerer) [inline api.php:1177]: I see, so this triggers the pusher to send messages? Very cool!

I think you need to move this inside of `libGameMessage::send` to catch messages sent from places other than the beta UI. Even then, I'm not sure it catches everything... you should grep for inserts into `wD_Messages` to see if there are any other places.

**Review by adamlerer — CHANGES_REQUESTED** [review]
So this actually works? That's amazing that you could switch this to websockets with so few changes. It would be great to switch the game overview fetching to web sockets as well, although figuring out the triggers might be harder there (maybe you can just catch gamemaster process events, and then augment that by an infrequent poll).

File: composer.lock (hunk 2 above)
Comment (utiq) [reply composer.lock:7]: check the composer.json file. I modified it to add the pusher library. So the compose.lock file updated its hash

**Comment by utiq** [comment]
> So this actually works? That's amazing that you could switch this to websockets with so few changes. It would be great to switch the game overview fetching to web sockets as well, although figuring out the triggers might be harder there (maybe you can just catch gamemaster process events, and then augment that by an infrequent poll).

Yeah, and the best part about it is that we can use open-source solutions or paid solutions.
For the gameoverview endpoint updates, I only added the WebSocket trigger in the gamemaster.php file. Where else do you think it's missing?

File: api.php (hunk 3 above)
Comment (lightvector) [reply api.php:1177]: It looks to me like libGameMessage is the central point for message sending, a bit of grepping didn't obviously turn up other places where it gets modified.

File: config.sample.php [hunk 4]
```diff
@@ -475,6 +475,47 @@ public static function isOnPlayNowDomain()
 		}
 		return false;	
 	}
-}
 
+	// ---
```
Comment (lightvector) [inline config.sample.php:479]: Are the various keys and secrets here something that we will have to note for the merge and deploy? e.g. they are something Kestas will need to privately pick some secure values for the production version of the config upon merge and deploy of a version of webdip after this change is merged in?

**Comment by lightvector** [comment]
> For the gameoverview endpoint updates, I only added the WebSocket trigger in the gamemaster.php file. Where else do you think it's missing?

GameOverviewResponse unfortunately is a catch-all object for practically every piece of data about a game that doesn't relate to the configuration of pieces on the actual board. See https://github.com/adamlerer/webDiplomacy/blob/main/beta-src/src/state/interfaces/GameOverviewResponse.ts for all the fields it has, most of which we use.

A lot of the fields of a GameOverviewResponse should stay constant after a game gets created, barring unusual mod intervention (there are many things a site moderator or admin can do to alter properties of a game that's started). Updating on process is the big one of course, but I think we also want to recognize:

* When users join/leave the game, such as due to civil disorder and then a new player joining and replacing the old one. 
* When the game finishes (to the degree that these are separate events from processing a game - for example if players agree to a draw, that handled as a game processing the phase the same way as if they all ready orders, or is that a separate thing?).
* When the game is paused or resumed or forcibly stopped/canceled (also to the degree that these are separate events from processing a game).

One of the fields is a little bit concerningly named: "pauseTimeRemaining". If we use that field anywhere, then we should make sure that "time remaining" would be handled correctly if we're going to switch away from polling and potentially going long stretches without updating, and/or instead change it to describe the end time of the pause or else something that isn't query-time-dependent.

If we're concerned about rare changes to other game metadata reflecting properly in player's UIs and it's hard to track down every last spot that it could happen, then maybe we could push websocket for just the common ones, and then still have a polling loop in addition to that to, where the polling loop runs very infrequently.

**Comment by adamlerer** [comment]
@lightvector yeah, I think we should add a gameOverview poll every 20s or something or we're going to have an infinite long tail of corner cases we have to add triggers for.

File: config.sample.php (hunk 4 above)
Comment (utiq) [reply config.sample.php:479]: Good point. Do you know if he manages that using ENV variables or some other ENV variables manager?

Comment (lightvector) [reply config.sample.php:479]: No idea. Just a thing to note somewhere so that we don't forget when it comes time to do the next merge and release.

**Comment by adamlerer** [comment]
I think we might need authentication before this can get merged... what do you think @lightvector ? It's true we're not sending the messages in the push, but just knowing who is receiving messages in an arbitrary game seems problematic...

**Comment by utiq** [comment]
> I think we might need authentication before this can get merged... what do you think @lightvector ? It's true we're not sending the messages in the push, but just knowing who is receiving messages in an arbitrary game seems slightly problematic...
> 
> If authentication is hard, we could ask Kestas what he thinks... I doubt there are many people trying to hack webdip to cheat.

it's not too hard, but it will take time. I would like to work on more tasks that are in the backlog.
What I could do is hide the country id and just send the signal with the game identifier.

**Comment by adamlerer** [comment]
@lightvector 

> One of the fields is a little bit concerningly named: "pauseTimeRemaining". If we use that field anywhere, then we should make sure that "time remaining" would be handled correctly if we're going to switch away from polling and potentially going long stretches without updating, and/or instead change it to describe the end time of the pause or else something that isn't query-time-dependent.

I was also confused about this field, but what it actually means is "how much time is remaining in this phase *after the pause ends*". So this does not change over time as long as the pause is still active. (This does mean we have to trigger when a pause toggles... or do the 20s poll thing).

**Comment by utiq** [comment]
do we merge it into adamlerer:main first and then I open a new PR to kestasjk:staging?
Or I edit this PR pointing to kestasjk:staging?

**Pull request merged**